tests/sbd-testbed.c

```diff
@@ -0,0 +1,746 @@
+            watchdog_arm();
+        }
+    }
+
+    return orig_write(fd, buf, count);
+}
+
+int
+ioctl(int fd, unsigned long int request, ...)
+{
+    int rv = -1;
+    va_list ap;
+    int i;
+
+    init();
+
+    va_start(ap, request);
+    switch (request) {
+        case BLKSSZGET:
+            for (i=0; i < 3; i++) {
+                if (sbd_device_fd[i] == fd) {
+                    rv = 0;
+                    *(va_arg(ap, int *)) = 512;
+                    break;
+                }
+                if (i == 2) {
+                    rv = orig_ioctl(fd, request, va_arg(ap, int *));
+                }
+            }
+            break;
+        case WDIOC_SETTIMEOUT:
+            if (fd == watchdog_device_fd) {
+                watchdog_timeout = *va_arg(ap, int *);
+
+                watchdog_arm();
+                rv = 0;
+                break;
+            }
+            rv = orig_ioctl(fd, request, va_arg(ap, int *));
+            break;
+        case WDIOC_SETOPTIONS:
+            if (fd == watchdog_device_fd) {
+                int flags = *va_arg(ap, int *);
+
+                if (flags & WDIOS_DISABLECARD) {
+                    watchdog_disarm();
+                }
+                rv = 0;
+                break;
+            }
+            rv = orig_ioctl(fd, request, va_arg(ap, int *));
+            break;
+        case WDIOC_GETSUPPORT:
+            rv = orig_ioctl(fd, request, va_arg(ap, struct watchdog_info *));
+            break;
+        default:
+            fprintf(log_fp, "ioctl using unknown request = 0x%08lx", request);
+            rv = orig_ioctl(fd, request, va_arg(ap, void *));
+    }
+    va_end(ap);
+
+    return rv;
+}
+
+int
+close(int fd)
+{
+    int i;
+
+    init();
+
+    if (fd == watchdog_device_fd) {
+        watchdog_device_fd = -1;
+    } else {
+        for (i = 0; i < 3; i++) {
+            if (sbd_device_fd[i] == fd) {
+                sbd_device_fd[i] = -1;
+                break;
+            }
+        }
+    }
+    return orig_close(fd);
+}
+
+// ***** end - handling of watchdog & block-devices ****
+
+// ***** handling of sysrq, sysrq-trigger & reboot ****
+
+FILE *
+fopen(const char *pathname, const char *mode)
+{
+    int is_sysrq = 0;
+    int is_sysrq_trigger = 0;
+    FILE *fp;
+
+    init();
+
+    if ((strcmp("/proc/sys/kernel/sysrq", pathname) == 0) &&
+        strcmp("w", mode)) {
+        pathname = "/dev/null";
+        is_sysrq = 1;
+    } else if (strcmp("/proc/sysrq-trigger", pathname) == 0) {
+        pathname = "/dev/null";
+        is_sysrq_trigger = 1;
+    }
+    fp = orig_fopen(pathname, mode);
+    if (is_sysrq) {
+        sysrq_fp = fp;
+    } else if (is_sysrq_trigger) {
+        sysrq_trigger_fp = fp;
+    }
+    return fp;
+}
+
+int
+fclose(FILE *fp)
+{
+    init();
+
+    if (fp == sysrq_fp) {
+        sysrq_fp = NULL;
+    } else if (fp == sysrq_trigger_fp) {
+        sysrq_trigger_fp = NULL;
+    }
+    return orig_fclose(fp);
+}
+
+#if __USE_FORTIFY_LEVEL > 1
```

Comment:
maybe "defined(__USE_FORTIFY_LEVEL) &&" for non-gcc compilers?

Reply:
That is a bit hacky.
Probably cleaner checking if it is defined at all although stdio2.h has the same clause.
As long as stdio2.h is looking as it does currently and if __USE_FORtIFY_LEVEL is not > 1
the mapping from fprintf to fprintf_chk will not happen and fprintf is directly used both here
and in the code that is actually using fprintf. If a similar logic is introduced for compilers
using something else the build is probably gonna fail. If the alternate approach isn't
breaking the build interception of sysrq_trigger access isn't gonna work and given
the user executing the test has the proper rights we might see an unwanted reboot.